File: LeetCode - Algorithm/Array - String/274.js
/**
 * @param {number[]} citations
 * @return {number}
 */
var hIndex = function (citations) {
  const sortedCitations = citations.sort((a, b) => b - a);

  result = 0;

  for (let i = 0; i < sortedCitations.length; i++) {
    if (sortedCitations[i] > i + 1) result = Math.max(result, i + 1);
    else {
      result = Math.max(result, sortedCitations[i]);

      break;
    }
  }

  return result;
};

`
[6,5,3,1,0]
 ^

[3,1,1]
 ^

[10, 10, 10]
 ^
`;

`
I believe I performed very well this time. 

Initially, I thought I would update the result whenever the value became equal to or smaller than the index. 
For example, given the array [5, 4, 3, 2, 1], I would stop at the 3rd number because the value 3 is equal to
the index 3 (i + 1). However, this approach does not work for cases like [10, 10, 10], where the index needs
to be considered to update the result correctly.
`;

File: LeetCode - Algorithm/Array - String/274.py
# 274. H-Index
# https://leetcode.com/problems/h-index/

from typing import List


class Solution:
    def hIndex(self, citations: List[int]) -> int:
        sortedCitations = sorted(citations)

        length = len(citations)
        index = 0
        while length - index > 0:
            if sortedCitations[length - 1 - index] > index:
                index += 1
            else:
                break

        return index


# index = 0
#
# 0, 1, 3, 5, 6
#             ^
